backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/blobs/CassandraBlobMapper.java

```diff
@@ -0,0 +1,138 @@
+import static com.datastax.driver.core.querybuilder.QueryBuilder.select;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.io.SequenceInputStream;
+import java.nio.ByteBuffer;
+import java.util.Enumeration;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.james.backends.cassandra.utils.CassandraAsyncExecutor;
+import org.apache.james.util.FluentFutureStream;
+
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.utils.UUIDs;
+import com.google.common.base.Preconditions;
+
+public class CassandraBlobMapper {
+
+
+    public static final int CHUNK_SIZE = 1024;
+    private final CassandraAsyncExecutor executor;
+
+    public CassandraBlobMapper(Session session) {
+        this.executor = new CassandraAsyncExecutor(session);
+    }
+
+    public CompletableFuture<UUID> write(byte[] data) {
+        UUID uuid = UUIDs.timeBased();
+        return FluentFutureStream.of(
+            computeBlobParts(data)
+                .map(pair ->
+                    executor.executeVoid(insertInto(CassandraBlobModule.TABLE_NAME)
+                        .value(CassandraBlobModule.ID, uuid)
+                        .value(CassandraBlobModule.PART, pair.getKey())
+                        .value(CassandraBlobModule.DATA, pair.getValue()))))
+            .completableFuture()
+        .thenApply(any -> uuid);
+    }
+
+    private Stream<Pair<Integer, ByteBuffer>> computeBlobParts(byte[] data) {
+        int size = data.length;
+        int fullChunkCount = size / CHUNK_SIZE;
+
+        return Stream.concat(
+            IntStream.range(0, fullChunkCount)
+                .mapToObj(i -> Pair.of(i, ByteBuffer.wrap(data, i * CHUNK_SIZE, CHUNK_SIZE))),
+            computeFinalByteBuffer(data, fullChunkCount * CHUNK_SIZE, fullChunkCount));
+    }
+
+    private Stream<Pair<Integer, ByteBuffer>> computeFinalByteBuffer(byte[] data, int offset, int index) {
+        if (offset == data.length) {
+            return Stream.of();
+        }
+        return Stream.of(Pair.of(index, ByteBuffer.wrap(data, offset, data.length - offset)));
+    }
+
+    public InputStream read(UUID id) {
+        return new SequenceInputStream(
+            new CassandraBackedInputStreamEnumeration(
+                i -> readPart(id, i)));
+    }
+
+    public CompletableFuture<Optional<Row>> readPart(UUID uuid, int part) {
+        return executor.executeSingleRow(select().from(CassandraBlobModule.TABLE_NAME)
+            .where(eq(CassandraBlobModule.ID, uuid))
+            .and(eq(CassandraBlobModule.PART, part)));
+    }
+
+    public static class CassandraBackedInputStreamEnumeration implements Enumeration<InputStream> {
```

Comment:
No choice: sequence InputStream expects an enumeration. Thats the only way to implement it with this class and I did not find equivalents